### PR TITLE
feat: merge mowali branch

### PR DESCRIPTION
**Pull request by shashi165**
Summary of changes
1. Now users can provide additional non-standard headers via rules.json. This will enable users to provide additional headers while redirecting the quote request
3. updated the documentation
4. some refactoring of code

### CODE REVIEW

File: src/interface/QuotingService-swagger.yaml [hunk 1]
```diff
@@ -1149,7 +1149,7 @@ components:
       type: object
       properties:
         extension:
-          maxItems: 16
+          maxItems: 17
```
Comment (elnyry-sam-k) [inline src/interface/QuotingService-swagger.yaml:1152]: @shashi165 - why this change to 17?

File: src/lib/util.js [hunk 2]
```diff
@@ -155,6 +155,12 @@ function generateRequestHeaders (headers, noAccept) {
   if (!noAccept) {
     ret.Accept = acceptHeader
   }
+  if (headers['fspiop-sourcecurrency']) {
```
Comment (elnyry-sam-k) [inline src/lib/util.js:158]: oooh

Comment (elnyry-sam-k) [inline src/lib/util.js:158]: lets discuss these

File: src/model/quotes.js [hunk 3]
```diff
@@ -131,14 +122,21 @@ class QuotesModel {
...
         terminate: false,
         quoteRequest,
         headers: {
           ...headers,
           'fspiop-destination': interceptQuoteEvents[0].params.rerouteToFsp
         }
       }
+      // if rerouteToFspCurrency is present then add the non standard headers used by forex
+      // Note these headers are not part of the mojaloop specification
+      if (interceptQuoteEvents[0].params.rerouteToFspCurrency) {
```
Comment (mdebarros) [inline src/model/quotes.js:135]: As discussed, please re-factor this to be generic, and rather have the rules provide "additionalHeaders" which can be injected.

File: src/model/quotes.js [hunk 4]
```diff
@@ -164,8 +162,9 @@ class QuotesModel {
       throw ErrorHandler.CreateInternalServerFSPIOPError('Missing quoteRequest', null, fspiopSource)
     }
 
-    await this.db.getParticipant(fspiopSource, LOCAL_ENUM.PAYER_DFSP, quoteRequest.amount.currency, ENUM.Accounts.LedgerAccountType.POSITION)
-    await this.db.getParticipant(fspiopDestination, LOCAL_ENUM.PAYEE_DFSP, quoteRequest.amount.currency, ENUM.Accounts.LedgerAccountType.POSITION)
+    // disabled below validation because of FX transfers the currencies are different
+    // await this.db.getParticipant(fspiopSource, LOCAL_ENUM.PAYER_DFSP, quoteRequest.amount.currency, ENUM.Accounts.LedgerAccountType.POSITION)
```
Comment (mdebarros) [inline src/model/quotes.js:166]: Please remove commented code. If you think we need to keep it, please include a reason as to why it's here.

This comment applies throughout this PR.

File: src/model/rules.js [hunk 5]
```diff
@@ -65,6 +65,15 @@ const createEngine = () => {
   engine.addOperator('deepEqual', (factValue, ruleValue) => {
     return deepEqual(factValue, ruleValue)
   })
+  engine.addOperator('isString', (factValue, ruleValue) => {
```
Comment (mdebarros) [inline src/model/rules.js:68]: Is it possible to add documentation on how to use the rules for the quoting-service?

File: src/handlers/quotes.js [hunk 6]
```diff
@@ -76,15 +76,15 @@ module.exports = {
       }, EventSdk.AuditEventAction.start)
 
       // call the quote request handler in the model
-      model.handleQuoteRequest(request.headers, request.payload, span)
+      const result = await model.handleQuoteRequest(request.headers, request.payload, span)
```
Comment (mdebarros) [inline src/handlers/quotes.js:79]: Major issues that I see here:

1. Any reason we can't split out the rules engine as a separate sync response?
2. Why are we not handling rule failures (I.e. business rules) by the expected FPSIOP Error Callback instead of handling this sync? <-- I think this will need to be addressed by the Design Authority.

@elnyry-sam-k, let me know your thoughts here.

**Review by mdebarros — CHANGES_REQUESTED** [review]
See my comments.

File: src/handlers/quotes.js (hunk 6 above)
Comment (lewisdaly) [reply src/handlers/quotes.js:79]: +1 @mdebarros - we shouldn't use `await` here, since we want to reply to the DFSP with a `202` as soon as possible

File: src/handlers/quotes/{id}/error.js [hunk 7]
```diff
@@ -74,14 +74,14 @@ module.exports = {
         payload: request.payload
       }, EventSdk.AuditEventAction.start)
       // call the quote error handler in the model
-      model.handleQuoteError(request.headers, quoteId, request.payload.errorInformation, span)
+      const result = await model.handleQuoteError(request.headers, quoteId, request.payload.errorInformation, span)
```
Comment (lewisdaly) [inline src/handlers/quotes/{id}/error.js:77]: Same as above, why add `await`?

File: src/model/quotes.js [hunk 8]
```diff
@@ -66,32 +68,21 @@ class QuotesModel {
...
-      axios.request({ url: `${url}/${headers['fspiop-source']}` }),
-      axios.request({ url: `${url}/${headers['fspiop-destination']}` })
-    ])
-
-    this.writeLog(`Got rules engine facts payer ${payer} and payee ${payee}`)
-
     const facts = {
-      payer,
-      payee,
+      payer: payer.data,
```
Comment (lewisdaly) [inline src/model/quotes.js:77]: Why `payer.data` and `payee.data` here? It strikes me that `fetchParticipantInfo` should return the payer and payee, not the axios response objects.

File: src/model/quotes.js (hunk 3 above)
Comment (lewisdaly) [reply src/model/quotes.js:135]: At the very least, this type of functionality should be behind a feature switch.

File: src/model/quotes.js [hunk 9]
```diff
@@ -178,6 +177,34 @@ class QuotesModel {
         await this.db.getParticipant(quoteRequest.payee.partyIdInfo.fspId, LOCAL_ENUM.PAYEE_DFSP, quoteRequest.amount.currency, ENUM.Accounts.LedgerAccountType.POSITION)
       }
     }
+
+    // check if the payer is active fsp, if not send error callback to payer
+    if (payer.data.isActive === 0) {
```
Comment (lewisdaly) [inline src/model/quotes.js:182]: Same as above, why `payer.data`?

File: src/model/quotes.js [hunk 10]
```diff
@@ -178,6 +177,34 @@ class QuotesModel {
...
+      throw ErrorHandler.CreateFSPIOPError(ErrorHandler.Enums.FSPIOPErrorCodes.PAYER_FSP_ID_NOT_FOUND,
+          `Payer FSP ID not found - Unsupported participant '${fspiopSource}'`, null, fspiopSource)
+    }
+    // check if the payee is active fsp, if not send error callback to payer
+    if (payee.data.isActive === 0) {
+      throw ErrorHandler.CreateFSPIOPError(ErrorHandler.Enums.FSPIOPErrorCodes.DESTINATION_FSP_ERROR,
+          `Destination FSP Error - '${fspiopDestination}' is inactive`, null, fspiopSource)
+    }
+
+    const payerAccounts = Array.isArray(payer.data.accounts) ? payer.data.accounts : []
```
Comment (lewisdaly) [inline src/model/quotes.js:192]: This sort of validation should be handled by `fetchParticipantInfo` for a cleaner separation of oncerns

File: src/model/quotes.js [hunk 11]
```diff
@@ -389,13 +419,14 @@ class QuotesModel {
       // lookup payee dfsp callback endpoint
       // TODO: for MVP we assume initiator is always payer dfsp! this may not always be the
       // case if a xfer is requested by payee
-      if (envConfig.simpleRoutingMode) {
-        endpoint = await this.db.getParticipantEndpoint(fspiopDest, 'FSPIOP_CALLBACK_URL_QUOTES')
-      } else {
-        endpoint = await this.db.getQuotePartyEndpoint(quoteId, 'FSPIOP_CALLBACK_URL_QUOTES', 'PAYEE')
-      }
+      // if (envConfig.simpleRoutingMode) {
```
Comment (lewisdaly) [inline src/model/quotes.js:422]: Why is this commented out?

File: src/model/quotes.js [hunk 12]
```diff
@@ -632,12 +663,13 @@ class QuotesModel {
       }
 
       // lookup payer dfsp callback endpoint
-      if (envConfig.simpleRoutingMode) {
-        endpoint = await this.db.getParticipantEndpoint(fspiopDest, 'FSPIOP_CALLBACK_URL_QUOTES')
-      } else {
-        // todo: for MVP we assume initiator is always payer dfsp! this may not always be the case if a xfer is requested by payee
-        endpoint = await this.db.getQuotePartyEndpoint(quoteId, 'FSPIOP_CALLBACK_URL_QUOTES', 'PAYER')
-      }
+      // if (envConfig.simpleRoutingMode) {
```
Comment (lewisdaly) [inline src/model/quotes.js:666]: Why is this commented out?

File: src/model/quotes.js [hunk 13]
```diff
@@ -1088,6 +1138,18 @@ class QuotesModel {
   writeLog (message) {
     Logger.info(`${new Date().toISOString()}, (${this.requestId}) [quotesmodel]: ${message}`)
   }
+
+  async fetchParticipantInfo (source, destination) {
```
Comment (lewisdaly) [inline src/model/quotes.js:1142]: I think this should live somewhere like `/src/lib`

**Review by lewisdaly — CHANGES_REQUESTED** [review]

File: src/interface/QuotingService-swagger.yaml (hunk 1 above)
Comment (shashi165) [reply src/interface/QuotingService-swagger.yaml:1152]: @elnyry-sam-k , Mowali has 17 elements, we need to check why they have 1 more (I am not making the change myself. I am just trying to merge whatever is there, I might not be able to answer why it was changed as it was changed long time ago)

File: src/lib/util.js (hunk 2 above)
Comment (shashi165) [reply src/lib/util.js:158]: sure

File: src/handlers/quotes.js (hunk 6 above)
Comment (shashi165) [reply src/handlers/quotes.js:79]: We need to discuss this further as per discussion with Miguel

File: src/model/quotes.js (hunk 3 above)
Comment (mdebarros) [reply src/model/quotes.js]: @lewisdaly that was also discussed.

File: src/interface/QuotingService-swagger.yaml (hunk 1 above)
Comment (mdebarros) [reply src/interface/QuotingService-swagger.yaml:1152]: @shashi165, someone will need to log a Change Request to the FSPIOP Change Control Board for this change with the appropriate rationale.

**Comment by lewisdaly** [comment]
> I'm concerned there are no integration tests for this new functionality. Can you please write an integration test that shows the new rules all working together?

@shashi165 any response to my earlier request?

**Review by mdebarros — DISMISSED** [review]
+1

**Comment by lewisdaly** [comment]
> > > I'm concerned there are no integration tests for this new functionality. Can you please write an integration test that shows the new rules all working together?
> > 
> > 
> > @shashi165 any response to my earlier request?
> 
> Hey @lewisdaly , there are no Integration Tests for this service, i.e. --> https://github.com/mojaloop/quoting-service/tree/master/test
> 
> The best we can do at this stage is to add a backlog item.

That's troubling. @shashi165 can you please open a backlog item to add integration tests to this service as a part of this task?

**Comment by shashi165** [comment]
> > > > I'm concerned there are no integration tests for this new functionality. Can you please write an integration test that shows the new rules all working together?
> > > 
> > > 
> > > @shashi165 any response to my earlier request?
> > 
> > 
> > Hey @lewisdaly , there are no Integration Tests for this service, i.e. --> https://github.com/mojaloop/quoting-service/tree/master/test
> > The best we can do at this stage is to add a backlog item.
> 
> That's troubling. @shashi165 can you please open a backlog item to add integration tests to this service as a part of this task?

Hi @lewisdaly - created a story here: https://github.com/mojaloop/project/issues/2570

File: package.json [hunk 14]
```diff
@@ -2,7 +2,7 @@
   "name": "quoting-service",
   "description": "Quoting Service hosted by a scheme",
   "license": "Apache-2.0",
-  "version": "13.0.1",
+  "version": "13.1.0",
```
[the remark this answers is not in the document]
Comment (shashi165) [reply package.json]: reverted the version

**Comment by mdebarros** [comment]
> Hey @shashi165,
> 
> I don't see any issues with the PR. But perhaps we should build a snapshot release and test it against the GP TTK tests before doing a final merge.
> 
> Let me know your thoughts on that. I would be happy to schedule a 1 hour call to do so....
> 
> Also see my one issue that you will need to address around changing the package version as it is already handled by our automated CI-CD releases.

[GP Report](https://mojaloop-oss-qa-results.s3.us-west-2.amazonaws.com/ttk-tests/reports/TTK-Assertion-Report-multi-2022-02-09T14%3A30%3A54.590Z.html)
```
--------------------FINAL REPORT--------------------
Test Suite:GP Tests
Environment:moja2.test.mojaloop.live
┌───────────────────────────────────────────────────┐
│                      SUMMARY                      │
├───────────────────┬───────────────────────────────┤
│ Total assertions  │ 2318                          │
├───────────────────┼───────────────────────────────┤
│ Passed assertions │ 2311                          │
├───────────────────┼───────────────────────────────┤
│ Failed assertions │ 7                             │
├───────────────────┼───────────────────────────────┤
│ Total requests    │ 551                           │
├───────────────────┼───────────────────────────────┤
│ Total test cases  │ 120                           │
├───────────────────┼───────────────────────────────┤
│ Passed percentage │ 99.70%                        │
├───────────────────┼───────────────────────────────┤
│ Started time      │ Wed, 09 Feb 2022 14:28:10 GMT │
├───────────────────┼───────────────────────────────┤
│ Completed time    │ Wed, 09 Feb 2022 14:30:54 GMT │
├───────────────────┼───────────────────────────────┤
│ Runtime duration  │ 164301 ms                     │
└───────────────────┴───────────────────────────────┘
```

[Bulk Report](https://mojaloop-oss-qa-results.s3.us-west-2.amazonaws.com/ttk-tests/reports/TTK-Assertion-Report-multi-2022-02-09T14%3A31%3A49.978Z.html)
```
--------------------FINAL REPORT--------------------
Test Suite:Bulk Tests
Environment:moja2.test.mojaloop.live
┌───────────────────────────────────────────────────┐
│                      SUMMARY                      │
├───────────────────┬───────────────────────────────┤
│ Total assertions  │ 134                           │
├───────────────────┼───────────────────────────────┤
│ Passed assertions │ 25                            │
├───────────────────┼───────────────────────────────┤
│ Failed assertions │ 109                           │
├───────────────────┼───────────────────────────────┤
│ Total requests    │ 14                            │
├───────────────────┼───────────────────────────────┤
│ Total test cases  │ 3                             │
├───────────────────┼───────────────────────────────┤
│ Passed percentage │ 18.66%                        │
├───────────────────┼───────────────────────────────┤
│ Started time      │ Wed, 09 Feb 2022 14:31:07 GMT │
├───────────────────┼───────────────────────────────┤
│ Completed time    │ Wed, 09 Feb 2022 14:31:49 GMT │
├───────────────────┼───────────────────────────────┤
│ Runtime duration  │ 42433 ms                      │
└───────────────────┴───────────────────────────────┘
```

**Comment by shashi165** [comment]
Screen Shot 2022-02-09 at 14 14 20
test results of GP

**Review by mdebarros — APPROVED** [review]

**Comment by mdebarros** [comment]
+1

```
Test Suite:GP Tests
Environment:Development
┌───────────────────────────────────────────────────┐
│                      SUMMARY                      │
├───────────────────┬───────────────────────────────┤
│ Total assertions  │ 2318                          │
├───────────────────┼───────────────────────────────┤
│ Passed assertions │ 2318                          │
├───────────────────┼───────────────────────────────┤
│ Failed assertions │ 0                             │
├───────────────────┼───────────────────────────────┤
│ Total requests    │ 551                           │
├───────────────────┼───────────────────────────────┤
│ Total test cases  │ 120                           │
├───────────────────┼───────────────────────────────┤
│ Passed percentage │ 100.00%                       │
├───────────────────┼───────────────────────────────┤
│ Started time      │ Thu, 10 Feb 2022 18:44:35 GMT │
├───────────────────┼───────────────────────────────┤
│ Completed time    │ Thu, 10 Feb 2022 18:47:06 GMT │
├───────────────────┼───────────────────────────────┤
│ Runtime duration  │ 151522 ms                     │
└───────────────────┴───────────────────────────────┘
TTK-Assertion-Report-multi-2022-02-10T18:47:06.854Z.html was generated
```


```
Test Suite:Bulk Tests
Environment:Development
┌───────────────────────────────────────────────────┐
│                      SUMMARY                      │
├───────────────────┬───────────────────────────────┤
│ Total assertions  │ 134                           │
├───────────────────┼───────────────────────────────┤
│ Passed assertions │ 134                           │
├───────────────────┼───────────────────────────────┤
│ Failed assertions │ 0                             │
├───────────────────┼───────────────────────────────┤
│ Total requests    │ 14                            │
├───────────────────┼───────────────────────────────┤
│ Total test cases  │ 3                             │
├───────────────────┼───────────────────────────────┤
│ Passed percentage │ 100.00%                       │
├───────────────────┼───────────────────────────────┤
│ Started time      │ Thu, 10 Feb 2022 18:47:22 GMT │
├───────────────────┼───────────────────────────────┤
│ Completed time    │ Thu, 10 Feb 2022 18:47:25 GMT │
├───────────────────┼───────────────────────────────┤
│ Runtime duration  │ 3169 ms                       │
└───────────────────┴───────────────────────────────┘
TTK-Assertion-Report-multi-2022-02-10T18:47:25.595Z.html was generated
```
